### PR TITLE
Spec changes for split init

### PR DESCRIPTION
**Pull request by mppf**
Spec changes to go along with PR #14564.

Reviewed by @vasslitvinov - thanks!

- [x] full local testing

### CODE REVIEW

File: doc/rst/language/spec.rst [hunk 1]
```diff
@@ -1652,25 +1671,31 @@ in :ref:`Constants`. Ref variables are discussed in
...
+:ref:`Split_Initialization`. If split initialization can be applied, the
+compiler will identify a later assignment statement and the
+right-hand-side of that statement will form the initialization
+expression. If the ``initialization-part`` is omitted and split
+initialization cannot be applied, then the variable will need to be
+initialized to a default value. Only `var` and `const` variable
+declarations can be initialized to a default value. Not all types have a
+default value. Default values are described in
+:ref:`Default_Values_For_Types`.
+
```
Comment (vasslitvinov) [inline doc/rst/language/spec.rst]: Adjust the following paragraph:
```
If the ``type-part`` is omitted or refers to a generic type,
the initialization expression is required and will ...
```
Then, add `concrete` in `determine the concrete type of the variable`.

Continuing from there:
```
If the ``type-part`` is present, the initialization expression
must be coercible to the specified type or, if it is generic,
to its instantiation.

File: doc/rst/language/spec.rst [hunk 2]
```diff
@@ -1680,6 +1705,105 @@ defined in :ref:`Multiple_Variable_Declarations`.
...
+When the ``initialization-part`` is present, it forms the initialization
+expression. For local variables, if the ``initialization-part`` is
+omitted, the compiler will search forward in the function for the first
+assignment statement(s) setting that variable that occur before the
+variable is otherwise used. It will search only within block declarations
+``{ }`` and conditionals. These assignment statements are called
+applicable assignment statements. The variable will be initialized at the
+applicable assignment instead of being assigned to. This feature is
+called split initialization.
```
Comment (vasslitvinov) [inline doc/rst/language/spec.rst]:
```
These assignment statement(s) are called
applicable assignment statements. They perform
initialization, not assignment, of that variable.
This feature is called split initialization.
```

File: doc/rst/language/spec.rst [hunk 3]
```diff
@@ -1680,6 +1705,105 @@ defined in :ref:`Multiple_Variable_Declarations`.
 Multiple variables can be grouped together using a tuple notation as
 described in :ref:`Variable_Declarations_in_a_Tuple`.
 
+.. _Split_Initialization:
+
+Split Initialization
+~~~~~~~~~~~~~~~~~~~~
+
+When the ``initialization-part`` is present, it forms the initialization
+expression. For local variables, if the ``initialization-part`` is
```
Comment (vasslitvinov) [inline doc/rst/language/spec.rst]: Replace the above two lines with:
```
If the ``initialization-part`` of a local variable declaration is omitted .....
```

File: doc/rst/language/spec.rst [hunk 4]
```diff
@@ -1680,6 +1705,105 @@ defined in :ref:`Multiple_Variable_Declarations`.
...
+
+   .. BLOCK-test-chapeloutput
+
+      4
+
+Split initialization does not apply:
+
+ * to module-level variables, fields, config variables, or ``extern``
+   variables.
+ * to variables that are referred to in nested functions
```
Comment (vasslitvinov) [inline doc/rst/language/spec.rst]: "nested functions" is an implementation restriction, which should not be in the spec.

Comment (vasslitvinov) [inline doc/rst/language/spec.rst]: In the following bullets, replace `to variables where` with simply `when` .

File: doc/rst/language/spec.rst [hunk 5]
```diff
@@ -1680,6 +1705,105 @@ defined in :ref:`Multiple_Variable_Declarations`.
...
 described in :ref:`Variable_Declarations_in_a_Tuple`.
 
+.. _Split_Initialization:
+
+Split Initialization
+~~~~~~~~~~~~~~~~~~~~
+
+When the ``initialization-part`` is present, it forms the initialization
+expression. For local variables, if the ``initialization-part`` is
+omitted, the compiler will search forward in the function for the first
```
Comment (vasslitvinov) [inline doc/rst/language/spec.rst]: `first` --> `earliest`

File: doc/rst/language/spec.rst [hunk 6]
```diff
@@ -1680,6 +1705,105 @@ defined in :ref:`Multiple_Variable_Declarations`.
...
+
+Split initialization does not apply:
+
+ * to module-level variables, fields, config variables, or ``extern``
+   variables.
+ * to variables that are referred to in nested functions
+ * to variables where an applicable assignment statement setting that
+   variable could not be identified
+ * to variables where the variable is used in some way before the
+   applicable assignment statement
```
Comment (vasslitvinov) [inline doc/rst/language/spec.rst]: `before the applicable` --> `before the earliest`
`statement` --> `statement(s)`

**Comment by mppf** [comment]
@vasslitvinov - I updating taking almost all of your suggestions. Could you have another look and let me know if you are satisfied? Thanks.

File: doc/rst/language/spec.rst [hunk 7]
```diff
@@ -1652,25 +1667,34 @@ in :ref:`Constants`. Ref variables are discussed in
...
+
+If the ``type-part`` is omitted or refers to a generic type, an
+initialization expression as described above is required. Note that such
+initialization expressions can be in later statements if
+:ref:`Split_Initialization` us used. When the ``type-part`` is omitted or
+generic, the type of the variable is inferred from the initialization
+expression using local type inference described
+in :ref:`Local_Type_Inference`. If the ``type-part`` is present, the
+initialization expression must be coercible to the specified type or, if
+th ``type-part`` is generic, to its instantiation.
```
Comment (vasslitvinov) [inline doc/rst/language/spec.rst]: Oops, orphan `th`.

**Pull request merged**